chain-signatures/node/src/indexer_eth.rs

```diff
@@ -920,87 +821,54 @@ async fn process_block(
+    helios_client: &Arc<EthereumClient>,
     mut requests_indexed: mpsc::Receiver<BlockAndRequests>,
-    mut finalized_epoch_rx: mpsc::Receiver<FinalizedEpoch>,
+    mut finalized_block_rx: mpsc::Receiver<u64>,
     sign_tx: mpsc::Sender<IndexedSignRequest>,
     node_near_account_id: AccountId,
 ) -> anyhow::Result<()> {
-    // We keep the last 2 finalized epochs to handle retried blocks that happen to be older
-    let mut finalized_epochs: Vec<FinalizedEpoch> = Vec::new();
+    let mut finalized_block_number: Option<u64> = None;
+
     loop {
-        while let Some(BlockAndRequests {
+        let Some(BlockAndRequests {
             block_number,
             block_hash,
             indexed_requests,
         }) = requests_indexed.recv().await
-        {
-            loop {
-                // Check if block is in any of the finalized epochs
-                let mut block_found = false;
-                for epoch in &finalized_epochs {
-                    if let Some(finalized_block_hash) = epoch.blocks.get(&block_number) {
-                        block_found = true;
-                        if *finalized_block_hash == block_hash {
-                            tracing::info!("Block {block_number} is finalized!");
-                            send_indexed_requests(
-                                indexed_requests.clone(),
-                                sign_tx.clone(),
-                                node_near_account_id.clone(),
-                            );
-                            break;
-                        } else {
-                            tracing::error!(
-                                "Block {block_number} hash mismatch: expected {block_hash:?}, got {finalized_block_hash:?}. Chain re-orged."
-                            );
-                            //TODO: handle the block reorg case
-                            break;
-                        }
-                    }
-                }
-
-                if block_found {
-                    break;
-                }
-
-                tracing::warn!(
-                    "Block number {block_number} with hash: {block_hash:?} not in finalized epochs. "
-                );
+        else {
+            return Err(anyhow::anyhow!("Failed to receive indexed requests"));
+        };
 
-                // Check if block is in history (older than our oldest epoch)
-                if !finalized_epochs.is_empty() {
-                    let oldest_epoch = &finalized_epochs[0];
-                    if block_number < oldest_epoch.start_block {
-                        tracing::error!("Block {block_number} is in history");
-                        //TODO: handle the block in history case which happens when a block is retried
-                        break;
-                    }
-                }
+        // Wait for finalized block if needed
+        while finalized_block_number.is_none() || block_number > finalized_block_number.unwrap() {
```

Comment:
the unwrap() makes me question things so probably this is better:
```suggestion
        while finalized_block_number.map_or(true, |n| block_number > n) {
```